### ISSUE TITLE
Flatbuffer investigations collector

### ISSUE THREAD

**Issue by adrianschultz** [OpenSimulationInterface/open-simulation-interface#691]
Please add your learnings and results here about your flatbuffer investigations, so that we can have a detailed discussion in the performance and packaging WG.

Pros:
- faster than protobuffers -> especially on sensor model side (de-serialization)

Cons:
- more complex code especially for serialization -> Are the improvement gains worth the refactoring effort for existing code?

General remarks:
- Are there other possibilties to acchieve better performance?
- Flexibility vs performance: Do we want to keep OSI as flexible as it is right now and life with performance limitations or not?

**Comment by ThomasNaderBMW** [comment]
Investigations on our side regarding the traffic participant called "OSI-Pedestrian".
Thanks to @FabianPfeufferCAP and @tbleher !!

[OsiPedestrian-Flatbuffer-Protobuf-Comparison.pdf](https://github.com/OpenSimulationInterface/open-simulation-interface/files/10236366/OsiPedestrian-Flatbuffer-Protobuf-Comparison.pdf)

**Comment by jdsika** [comment]
For the sake of completeness I am adding further links to reports here:
1) [AVL report](https://github.com/OpenSimulationInterface/osi-sensor-model-packaging/blob/flatbuffers_object_api/doc/protob_flatb_compare/Flatbuffers.pdf)
2) [Persival report](https://github.com/OpenSimulationInterface/open-simulation-interface/files/8540495/Report_OSI_Flatbuffers_Performance.pdf)

**Comment by PhRosenberger** [comment]
In addition, check out our [Application report](https://github.com/OpenSimulationInterface/open-simulation-interface/files/8540492/Report_OSI_Flatbuffers_Application.pdf)

**Comment by jdsika** [comment]
@pmai should we just create a PR that puts nested ENUM definitions at the beginning of the message definition? I do not think that this would break backwards compatibility?

![grafik](https://user-images.githubusercontent.com/11943343/213474263-75f895ad-f949-4fbe-bad9-c122a242096f.png)

**Comment by jdsika** [comment]
@PhRosenberger I think we should open a feature request in the Flatbuffer GitHub and ask for the feature? What do you think?

`Flutbuffers does not come with a deep copy function, so there is no easy way to copy table or entire buffers to a different buffer. With OSI it is currently common practice to copy the SensorView input of a sensor model to the SensorData output. As there is no standard function to do this with Flatbuffers, one has to be custom written.`

**Comment by PhRosenberger** [comment]
> @PhRosenberger I think we should open a feature request in the Flatbuffer GitHub and ask for the feature? What do you think?
> 
> `Flutbuffers does not come with a deep copy function, so there is no easy way to copy table or entire buffers to a different buffer. With OSI it is currently common practice to copy the SensorView input of a sensor model to the SensorData output. As there is no standard function to do this with Flatbuffers, one has to be custom written.`

Yes we can do that, but I am a bit disappointment by the response to our issue over there, so I am not sure if we can expect it to actually happen. If we get some feedback, it would be nice to discuss such a feature with the google team behind FlatBuffers.

**Comment by jdsika** [comment]
> > @PhRosenberger I think we should open a feature request in the Flatbuffer GitHub and ask for the feature? What do you think?
> > `Flutbuffers does not come with a deep copy function, so there is no easy way to copy table or entire buffers to a different buffer. With OSI it is currently common practice to copy the SensorView input of a sensor model to the SensorData output. As there is no standard function to do this with Flatbuffers, one has to be custom written.`
> 
> Yes we can do that, but I am a bit disappointment by the response to our issue over there, so I am not sure if we can expect it to actually happen. If we get some feedback, it would be nice to discuss such a feature with the google team behind FlatBuffers.

Yes, I can understand this feeling. Let's just keep going and do the feature request. If we all start commenting on it we might get things moving :)

**Comment by PhRosenberger** [comment]
> Yes, I can understand this feeling. Let's just keep going and do the feature request. If we all start commenting on it we might get things moving :)

I just created such an issue, so we will know their thoughts on this soon, hopefully: https://github.com/google/flatbuffers/issues/7798

**Comment by PhRosenberger** [comment]
We should discuss the concept of nested flatbuffers, e.g. nesting SensorView and then nesting GlobalGroundTruth inside of the SensorView, as suggested by https://github.com/google/flatbuffers/issues/7798#issuecomment-1401363501

**Comment by PhRosenberger** [comment]
Copying of stuff (e.g. GlobalGroundTruth or SensorViewConfiguration) is now moved to https://github.com/OpenSimulationInterface/open-simulation-interface/issues/703

**Comment by thomassedlmayer** [comment]
Benchmark results/presentation slides - Flatbuf struct vs. table vs. protobuf based on https://github.com/OpenSimulationInterface/osi-sensor-model-packaging/tree/feature/flatbuffers_examples

[presentation-flatbuf-comparison.pdf](https://github.com/OpenSimulationInterface/open-simulation-interface/files/11285154/presentation-flatbuf-comparison.pdf)

**Comment by PhRosenberger** [comment]
> @pmai should we just create a PR that puts nested ENUM definitions at the beginning of the message definition? I do not think that this would break backwards compatibility?
> 
> ![grafik](https://user-images.githubusercontent.com/11943343/213474263-75f895ad-f949-4fbe-bad9-c122a242096f.png)

@jdsika and @pmai Was this PR already created? - If not, I would start it now :-)

**Comment by PhRosenberger** [comment]
As discussed in the performance and packaging workgroup last week, we will now start a questionnaire for the whole OSI project to get an overview on the usage of OSI and if there is a need for switching to Flatbuffers or not.

**Comment by PhRosenberger** [comment]
As discussed in the CCB meeting today, @pmai is preparing a PR for this. Thank you!

**Comment by thomassedlmayer** [comment]
I finally managed to run the benchmark of the BMW OSI Pedestrian code (https://github.com/OpenSimulationInterface/open-simulation-interface/issues/691#issuecomment-1352900872) with more objects (up to 79.000). I noticed that the mutate function was used for initially filling the flatbuffer structs. According to the flatbuffer tutorial mutate is not recommended as the default way of setting flatbuffer data (see https://flatbuffers.dev/flatbuffers_guide_tutorial.html; section "Mutating FlatBuffers"). Though, it did not significantly improve performance when changing the following structure:

```
auto size3d = osi3::Dimension3d();
size3d.mutate_height(size.height());
size3d.mutate_width(size.width());
size3d.mutate_length(size.length());
```

to:

`auto size3d = osi3::Dimension3d(size.height(), size.width(), size.length());`

for all structs.

### Benchmark results:

![Figure_1](https://github.com/OpenSimulationInterface/open-simulation-interface/assets/65231631/8acba15e-3da3-469e-be2f-a77dac891124)

Zoomed in on max. 20.000 objects:

![Figure_2](https://github.com/OpenSimulationInterface/open-simulation-interface/assets/65231631/a9ea45ef-c197-4a70-8840-2f7294bb89c4)

Zoomed in on max. 10.000 objects:

![Figure_3](https://github.com/OpenSimulationInterface/open-simulation-interface/assets/65231631/68598097-5375-4770-b3db-1a3449c3bb96)

I did not run the benchmark multiple times as it was done for the initial BMW report, so I do not have any error/variance values. But the trend for up to 3000 objects is relatively similar to the BMW report. I think it should give a good enough general idea on the performance improvements.

**Comment by PhRosenberger** [comment]
Great to see these results @thomassedlmayer . Thanks for spending the efforts on reproducing it!
It looks way less linear than in the BMW report for the lower numbers of objects.

**Comment by florianwallerer** [comment]
As it was a question in todays CCB meeting (2023-09-11):
[Results - OSI Project Questionnaire - 23-06-12 - Protobuf vs. Flatbuffer](https://asamev.sharepoint.com/:x:/r/sites/osi/P_2022_01_ASAMOSI/03_development/Performance%20%26%20Packaging/Results%20Questionaire%20-%20Protobuf%20vs.%20Flatbuffer.xlsx?d=wc92a04de8a824ea29bf9f8c39bdbb3b4&csf=1&web=1&e=l6ZPgN)
